README.md
# Prometheus exporter for ElasticSearch
[![Build Status](https://travis-ci.org/monitoring-tools/prom-elasticsearch-exporter.svg?branch=master)](https://travis-ci.org/monitoring-tools/prom-elasticsearch-exporter)

### Installation

#### Pre-build binaries

You can find here: https://github.com/monitoring-tools/prom-elasticsearch-exporter/releases

#### Docker

```bash
docker pull montools/prom-elasticsearch-exporter
docker run --rm -p 9108:9108 montools/prom-elasticsearch-exporter:latest
```

### Build

#### 1. Clone repo

```bash
mkdir -p $GOPATH/src/github.com/monitoring-tools && cd $GOPATH/src/github.com/monitoring-tools
git clone https://github.com/monitoring-tools/prom-elasticsearch-exporter.git
cd prom-elasticsearch-exporter
```

#### 2. Build binary

```bash
make build
./prom-elasticsearch-exporter --es.uri="http://localhost:9200/"
```

#### 3. Build docker image

```bash
make docker
docker run -d -p 9108:9108 --name=prom-elasticsearch-exporter monitoring-tools/prom-elasticsearch-exporter:1.1.0 --es.uri="http://localhost:9200/"
```

### Configuration

```bash
prom-elasticsearch-exporter --help
```

| Argument              | Description |
| --------              | ----------- |
| web.listen-address    | Address to listen on for web interface and telemetry. Default - :9108 |
| web.telemetry-path    | Path under which to expose metrics. Default - /metrics |
| es.uri                | ElasticSearch URI. You can provide multiple hosts: --es.uri=host1 --es.uri=host2. If you're using multiple hosts and --es.all=true, metrics will be fetched from first responded node`.
| es.all                | If true - export stats for all nodes in the cluster. Default - false
| es.timeout            | Timeout for trying to get stats from ElasticSearch. Default - 5s |
| es.ca                 | Path to PEM file that contains trusted CAs for the Elasticsearch connection.
| es.client-private-key | Path to PEM file that contains the private key for client auth when connecting to Elasticsearch.
| es.client-cert        | Path to PEM file that contains the corresponding cert for the private key to connect to Elasticsearch.

### Grafana dashboards

To use this dashboards you need to set up following Prometheus [aggregation rules](examples/prometheus.rules).

Dashboards:
- [Cluster](examples/grafana-dashboard-cluster.json)
- [Index](examples/grafana-dashboard-index.json)
- [Node](examples/grafana-dashboard-node.json)

## License

This software is distributed under the terms of the [Apache License v2.0](LICENSE)

## Contributing

Contributions are very welcome.
Please fork the project on GitHub and open pull requests for any proposed changes.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.2.2] - 2020-01-05
### Changed
- Metric "tasks" was renamed to "task_group_duration_seconds".
### Added
- Label "cluster" to "task_group_duration_seconds" metric.

## [1.2.1] - 2019-12-24
### Added
- Nodes and indices requests cache hit and miss count 

## [1.2.0] - 2019-11-26
### Added
- Tasks running time metrics added 

## [1.1.0] - 2017-11-16
### Added
- CHANGELOG.md file was added

### Changed
- Multiple ElasticSearch nodes support removed.
  Now one exporter can fetch data from one ElasticSearch node.


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	_ "net/http/pprof"
	"os"
	"strings"
	"time"

	"github.com/monitoring-tools/prom-elasticsearch-exporter/collector"
	"github.com/monitoring-tools/prom-elasticsearch-exporter/elasticsearch"
	"github.com/monitoring-tools/prom-elasticsearch-exporter/encryption"
	"github.com/monitoring-tools/prom-elasticsearch-exporter/httpclient"
	"github.com/monitoring-tools/prom-elasticsearch-exporter/httpclient/decorator"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

const usage = `prom-elasticsearch-exporter - tool for exporting ElasticSearch metrics for Prometheus

Usage:

  prom-elasticsearch-exporter [commands|flags]

The commands & flags are:

  version                   print the version to stdout

  --web.listen-address      address to listen on for web interface and telemetry. Default - :9108
  --web.telemetry-path      path under which to expose metrics. Default - /metrics
  --es.timeout              timeout for trying to get stats from ElasticSearch. Default - 5s
  --es.uri                  ElasticSearch node URI. Default - http://localhost:9200
  --es.all                  export stats for all nodes in the cluster. Default - false
  --es.ca                   path to PEM file that conains trusted CAs for the ElasticSearch connection
  --es.client-private-key   path to PEM file that conains the private key for client auth when connecting to ElasticSearch
  --es.client-cert          path to PEM file that conains the corresponding cert for the private key to connect to Elasticsearch
`

// Variables passed through ldflags
var (
	version   string
	goVersion string
	gitBranch string
)

func main() {
	var (
		listenAddress      = flag.String("web.listen-address", ":9108", "Address to listen on for web interface and telemetry")
		metricsPath        = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics")
		esTimeout          = flag.Duration("es.timeout", 5*time.Second, "Timeout for trying to get stats from ElasticSearch")
		esURI              = flag.String("es.uri", "http://localhost:9200", "HTTP API address of an Elasticsearch node")
		esAllNodes         = flag.Bool("es.all", false, "Export stats for all nodes in the cluster")
		esCA               = flag.String("es.ca", "", "Path to PEM file that conains trusted CAs for the ElasticSearch connection")
		esClientPrivateKey = flag.String("es.client-private-key", "", "Path to PEM file that conains the private key for client auth when connecting to ElasticSearch")
		esClientCert       = flag.String("es.client-cert", "", "Path to PEM file that conains the corresponding cert for the private key to connect to ElasticSearch")
	)

	flag.Usage = func() { printUsage() }
	flag.Parse()

	args := flag.Args()

	if len(args) > 0 {
		switch args[0] {
		case "version":
			if version == "" {
				println("n/a")
			}
			fmt.Println(version)
			return
		case "help":
			printUsage()
		}
	}

	// returns nil if not provided and falls back to simple TCP.
	tlsConfig := encryption.CreateTLSConfig(*esCA, *esClientCert, *esClientPrivateKey)

	httpClient := &http.Client{
		Timeout: *esTimeout,
		Transport: &http.Transport{
			TLSClientConfig: tlsConfig,
		},
	}

	decoratedClient := httpclient.Decorate(
		httpClient,
		decorator.BaseURLDecorator(*esURI),
		decorator.RecoverDecorator(), // better to place it last to recover panics from decorators too
	)

	prometheus.MustRegister(collector.NewCompositeCollector(
		elasticsearch.NewClient(decoratedClient),
		*esAllNodes,
		version, goVersion, gitBranch,
	))

	http.Handle(*metricsPath, promhttp.Handler())
	http.HandleFunc("/", IndexHandler(*metricsPath))

	log.Println("Listening on:", formatListenAddr(*listenAddress))

	if err := http.ListenAndServe(*listenAddress, nil); err != nil {
		log.Println("Unable to start HTTP server:", err)
	}
}

// IndexHandler returns a http handler with the correct metricsPath
func IndexHandler(metricsPath string) http.HandlerFunc {
	indexHTML := `
<html>
	<head>
		<title>Elasticsearch Exporter</title>
	</head>
	<body>
		<h1>Elasticsearch Exporter</h1>
		<p>
			<a href='%s'>Metrics</a>
		</p>
	</body>
</html>
`
	index := []byte(fmt.Sprintf(strings.TrimSpace(indexHTML), metricsPath))

	return func(w http.ResponseWriter, r *http.Request) {
		w.Write(index)
	}
}

// formatListenAddr returns formatted UNIX addr
func formatListenAddr(addr string) string {
	parts := strings.Split(addr, ":")
	if len(parts) == 2 && parts[0] == "" {
		addr = fmt.Sprintf("localhost:%s", parts[1])
	}
	return "http://" + addr
}

func printUsage() {
	fmt.Println(usage)
	os.Exit(0)
}
